### PR TITLE
HistogramData rollout: algorithms MuonAnalysis

### PR DESCRIPTION
**Pull request by AnthonyLim23**
## Description of work.
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:
MantidQt/CustomInterfaces/src/Muon/ALCBaselineModellingModel.cpp
MantidQt/CustomInterfaces/src/Muon/ALCBaselineModellingPresenter.cpp
MantidQt/CustomInterfaces/src/Muon/ALCDataLoadingPresenter.cpp
MantidQt/CustomInterfaces/src/Muon/ALCHelper.cpp
MantidQt/CustomInterfaces/src/Muon/ALCPeakFittingPresenter.cpp
MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp
MantidQt/CustomInterfaces/src/Muon/MuonAnalysisFitDataPresenter.cpp
MantidQt/CustomInterfaces/src/Muon/MuonAnalysisHelper.cpp
MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h
MantidQt/CustomInterfaces/test/ALCBaselineModellingPresenterTest.h
MantidQt/CustomInterfaces/test/ALCPeakFittingModelTest.h
MantidQt/CustomInterfaces/test/MuonAnalysisDataLoaderTest.h
MantidQt/CustomInterfaces/test/MuonAnalysisHelperTest.h
Performance tests were added for:
MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h (may not be needed)

The following algorithms have improved performance:

## To test.

Code review.

<Add any manual testing not covered by unit tests, especially GUIs>

Fixes https://github.com/mantidproject/mantid/issues/18758 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidQt/CustomInterfaces/src/Muon/ALCBaselineModellingModel.cpp [hunk 1]
```diff
@@ -86,6 +86,7 @@ void ALCBaselineModellingModel::disableUnwantedPoints(
   // disable list
   for (size_t i = 0; i < ws->blocksize(); ++i) {
     for (auto it = sections.begin(); it != sections.end(); ++it) {
+
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/src/Muon/ALCBaselineModellingModel.cpp]: ?

File: MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp [hunk 2]
```diff
@@ -1296,7 +1296,7 @@ void MuonAnalysis::inputFileChanged(const QStringList &files) {
 
   // Populate bin width info in Plot options
   double binWidth =
-      matrix_workspace->dataX(0)[1] - matrix_workspace->dataX(0)[0];
+      matrix_workspace->dataX(0)[1] - matrix_workspace->mutableX(0)[0];
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp]: `dataX` -> `x`
`mutableX` -> `x`

File: MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp [hunk 3]
```diff
@@ -1755,15 +1755,15 @@ QMap<QString, QString> MuonAnalysis::getPlotStyleParams(const QString &wsName) {
...
       MatrixWorkspace_sptr matrix_workspace =
           boost::dynamic_pointer_cast<MatrixWorkspace>(ws_ptr);
-      const Mantid::MantidVec &dataY = matrix_workspace->readY(0);
+      const auto &yData = matrix_workspace->y(0);
 
       if (min.isEmpty())
-        params["YAxisMin"] =
-            QString::number(*min_element(dataY.begin(), dataY.end()));
+        params["YAxisMin"] = QString::number(
+            *min_element(yData.rawData().begin(), yData.rawData().end()));
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp]: Why `rawData()`?

File: MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp [hunk 4]
```diff
@@ -1755,15 +1755,15 @@ QMap<QString, QString> MuonAnalysis::getPlotStyleParams(const QString &wsName) {
...
 
       if (max.isEmpty())
-        params["YAxisMax"] =
-            QString::number(*max_element(dataY.begin(), dataY.end()));
+        params["YAxisMax"] = QString::number(
+            *max_element(yData.rawData().begin(), yData.rawData().end()));
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/src/Muon/MuonAnalysis.cpp]: Why `rawData()`?

File: MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h [hunk 5]
```diff
@@ -144,4 +145,65 @@ class ALCBaselineModellingModelTest : public CxxTest::TestSuite {
...
+  void setUp() override {
+    m_model = new ALCBaselineModellingModel();
+    xTestData = {10.0, 1.0, 1.41, 10.0, 10.0, 1.73, 2.0, 2.5, 10.0};
+    yTestData = {100, 1, 2, 100, 100, 3, 4, 5, 100};
+    eTestData = {1, 2, 3, 4, 5, 6, 7, 8, 9};
+  }
+
+  void tearDown() override { delete m_model; }
+
+  void test_setData() {
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h]: What are the runtimes of `test_setData` and `test_fit`? The test data is so tiny that I suspect it is not meaningful?

File: MantidQt/CustomInterfaces/test/MuonAnalysisDataLoaderTest.h [hunk 6]
```diff
@@ -196,6 +196,7 @@ class MuonAnalysisDataLoaderTest : public CxxTest::TestSuite {
       // Check that each period has number of spectra = number of groups
       TS_ASSERT_EQUALS(matrixWS->getNumberHistograms(), grouping.groups.size());
       // Check that each period has been corrected for dead time
+
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/test/MuonAnalysisDataLoaderTest.h]: ?

File: MantidQt/CustomInterfaces/test/MuonAnalysisDataLoaderTest.h [hunk 7]
```diff
@@ -363,4 +364,4 @@ class MuonAnalysisDataLoaderTest : public CxxTest::TestSuite {
   }
 };
 
-#endif /* MANTIDQT_CUSTOMINTERFACES_MUONANALYSISDATALOADERTEST_H_ */
\ No newline at end of file
+#endif /* MANTIDQT_CUSTOMINTERFACES_MUONANALYSISDATALOADERTEST_H_ */
```
Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/test/MuonAnalysisDataLoaderTest.h]: ?

**Review by SimonHeybrock — COMMENTED** [review]
Looks mostly good, a few comments below:

File: MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h (hunk 5 above)
Comment (AnthonyLim23) [inline MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h]: Yes I think you are right. I can remove them if you would like

Comment (SimonHeybrock) [inline MantidQt/CustomInterfaces/test/ALCBaselineModellingModelTest.h]: Either that, or use a more sensible test data size.

**Comment by AnthonyLim23** [comment]
I think that should address all of your comments

**Comment by SimonHeybrock** [comment]
Latest code changes look good. :shipit:

**Pull request merged**